lib/rebaser-server/src/server/core_loop.rs

```diff
@@ -0,0 +1,190 @@
+        };
+
+        if let Err(err) = message.ack_with(AckKind::Progress).await {
+            error!(error = ?err, "could not ack with progress, going to continue anyway");
+        }
+
+        // Deserialize the message payload so that we can process it.
+        let request_message: RequestRebaseMessage =
+            match serde_json::from_slice(message.message.payload.to_vec().as_slice()) {
+                Ok(deserialized) => deserialized,
+                Err(err) => {
+                    error!(error = ?err, ?message, "failed to deserialize message payload");
+                    continue;
+                }
+            };
+
+        // Pull the reply subject off of the message.
+        let reply_subject = if let Some(headers) = &message.headers {
+            if let Some(value) = headers.get(REPLY_SUBJECT_HEADER_NAME.clone()) {
+                value.to_string()
+            } else {
+                // NOTE(nick): we may actually want to process the message anyway, but things would be super messed up
+                // at that point... because no one should be sending messages exterior to rebaser clients.
+                error!(
+                    ?message,
+                    "no reply subject found in headers, skipping messages because we cannot reply"
+                );
+                continue;
+            }
+        } else {
+            // NOTE(nick): we may actually want to process the message anyway, but things would be super messed up
+            // at that point... because no one should be sending messages exterior to rebaser clients.
+            error!(
+                ?message,
+                "no headers found, skipping message because we cannot reply"
+            );
+            continue;
+        };
+
+        let mut ctx = match ctx_builder.build_default().await {
+            Ok(ctx) => ctx,
+            Err(err) => {
+                // NOTE(nick): we may actually want to bubble up the failure here.
+                error!(error = ?err, "unable to build dal context");
+                continue;
+            }
+        };
+        ctx.update_visibility(Visibility::new_head(false));
+        ctx.update_tenancy(Tenancy::new(WorkspacePk::NONE));
+
+        let start = Instant::now();
+        perform_rebase_and_reply_infallible(&mut ctx, request_message, reply_subject).await;
+        info!("perform rebase and reply total: {:?}", start.elapsed());
+    }
+}
+
+async fn perform_rebase_and_reply_infallible(
+    ctx: &mut DalContext,
+    message: RequestRebaseMessage,
+    reply_subject: impl ToSubject,
+) {
+    let reply_subject = reply_subject.to_subject();
+
+    let reply_message = perform_rebase(ctx, message).await.unwrap_or_else(|err| {
+        error!(error = ?err, ?message, ?reply_subject, "performing rebase failed, attempting to reply");
+        ReplyRebaseMessage::Error {
+            message: err.to_string(),
+        }
+    });
+
+    match serde_json::to_vec(&reply_message) {
```

Comment:
Reply back to the requester using regular NATS via a reply mailbox. We will do this for errors too.